### DISCUSSION TITLE
actions before saving image

### DISCUSSION THREAD

**Discussion by Tauxxx** [area17/twill#2685] [General]
How can I add actions before saving an image to the media gallery? For example, automatically reduce the size, optimize the image or add a watermark

**Comment by ifox** [comment]
Hi @Tauxxx you can bind your own `MediaRepository` to be able to implement standard Twill hook functions like `afterSave`, or your own `MediaLibraryController`, if you want to intercept the uploaded file. See this post for reference on binding custom classes for internal Twill classes: https://github.com/area17/twill/discussions/1232.

**Comment by Tauxxx** [comment]
> Hi @Tauxxx you can bind your own `MediaRepository` to be able to implement standard Twill hook functions like `afterSave`, or your own `MediaLibraryController`, if you want to intercept the uploaded file. See this post for reference on binding custom classes for internal Twill classes: #1232.

thanks. my solution
1. create controller 
```
namespace App\Http\Controllers\Admin;

use A17\Twill\Http\Controllers\Admin\MediaLibraryController;
use Intervention\Image\Facades\Image;

class AppMediaLibraryController extends MediaLibraryController

...
    public function storeFile($request)
...
    private function shouldReplaceMedia($id)

```
2. register controller
```
use A17\Twill\Http\Controllers\Admin\MediaLibraryController;
use App\Http\Controllers\Admin\AppMediaLibraryController;

class AppServiceProvider extends ServiceProvider
{
    /**
     * Register any application services.
     */
    public function register(): void
    {
        $this->app->bind(MediaLibraryController::class, AppMediaLibraryController::class);
    }
```
